### PR TITLE
fix: NUMERIC was missing from default mappings

### PR DESCRIPTION
**Pull request by olavloite**
Adds `NUMERIC` to the default mappings of `DbType => SpannerDbType`, `SpannerDbType => DbType` and `ClrType => SpannerDbType`.

Two considerations:
1. The Spanner `NUMERIC` data type is mapped to `DbType.VarNumeric` and not `DbType.Decimal`, as the latter has a lower range and precision than the Spanner `NUMERIC` type.
2. The .NET `decimal` ClrType [is currently mapped](https://github.com/olavloite/google-cloud-dotnet/blob/243c9fb4f32cd9e1cd48af07a48962c38f8e831e/apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs#L281) to the Spanner `FLOAT64` data type. That is correct in the sense that both are floating point types. It does however not correspond with the default mapping for other database systems:
2.1. [SQL Server](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/linq/sql-clr-type-mapping#numeric-mapping) maps it to `DECIMAL(29,4)`
2.2. [PostgreSQL](https://www.npgsql.org/doc/types/basic.html) maps it to `numeric` (which is a fixed point exact type)
2.3. [MySQL ](https://www.devart.com/dotconnect/mysql/docs/datatypemapping.html)maps it to `decimal` (which is a fixed point exact type).

Point 2 might cause some surprises for users moving from other DBMS'es to Spanner, and I would prefer to change that mapping to `NUMERIC` as well. WDYT?

### CODE REVIEW

**Comment by jskeet** [comment]
Will have a look at the PR later today - I think it would be best to file a new issue for point 2 for further discussion. (My initial guess is that we'd want to make that an option somewhere - just changing it immediately would be potentially breaking.)

I don't currently see any tests in the PR... it would be great to see a test that currently fails, that's then fixed by the production change.

**Comment by olavloite** [comment]
> I don't currently see any tests in the PR... it would be great to see a test that currently fails, that's then fixed by the production change.

I added test cases for the change.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs:62]: Would it be worth it to add a method in SpannerDbType.FromDbType and move all the code in this set there? Just for the sake of having type conversions in one place?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs:59]: The literal suffixes are case insensitive, but you've used capital letters on the ones above, so, for consistency:
```suggestion
            yield return new object[] { (SpannerNumeric)3.14M, SpannerDbType.Numeric, DbType.VarNumeric };
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs:57]: Can you add a case for float and decimal as well which currently map to Float64? Actually, for all of these, be exhaustive.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
A few nits.
But, I suppose this isn't being used anywhere?? Or how did we manage to be without these mappings for SpannerNumeric before?
If it's being used somewhere, can you please at a test for that?
Or are we just using the conversions? If these mappings are really not being used, maybe we could consider depracating them?

**Comment by olavloite** [comment]
Thanks for the review. I'll try to look into your points later today.

> A few nits.
> But, I suppose this isn't being used anywhere?? Or how did we manage to be without these mappings for SpannerNumeric before?

These are only being used if the client application _only_ sets a _value_ and not a type. This is not very common when working specifically with the Spanner client library, but is more common when working with it through a framework that treats it like a generic `DbConnection`. So something like this would cause trouble:

```csharp
SpannerConnection connection = ...
var cmd = connection.CreateDbCommand();
var param = cmd.CreateDbParameter();
param.Value = (SpannerNumeric)3.14M;
```

When executing this command the Spanner client would try to infer the type from the value [here](https://github.com/googleapis/google-cloud-dotnet/blob/7adf800d20142a1ecd2e88d27fde290daf2e15b7/apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs#L121). In all other cases it would just use the type that the client application had supplied.

> If it's being used somewhere, can you please at a test for that?
I'll try to add some tests for using `SpannerCommand` with only a value and no specific type.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs:59]: I'd suggest the opposite, actually - I suggest we fix the suffixes above to be lower-case, as IMO that's more idiomatic C# :)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs:62]: +1 to this.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs:59]: I actually debated which was best, but got "confused" (is it lower case that I normally use, upper case?) and then I found [this](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types#integer-literals) (see the note there). Maybe lower case for anything except L? But I'm fine with the most idiomatic/consistent we can find ;).

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs
Comment (olavloite) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs:252]: While going through this class I also stumbled across this one. It is not in use anywhere and does not really make any sense.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs
Comment (olavloite) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs]: Done.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs
Comment (olavloite) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs]: Done (everything lower case except for long). I couldn't really find anything authoritative convention for this.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs
Comment (olavloite) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerParameterTests.cs]: I've added additional test cases for the currently existing mappings.

**Comment by olavloite** [comment]
> I'll try to add some tests for using `SpannerCommand` with only a value and no specific type.

I just realized that any test like that would just be a repetition of the `ValueMappings` test, which is literally doing this already.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs:212]:
```suggestion
        internal static SpannerDbType FromDbType(DbType dbType) => dbType switch
        {
            DbType.Binary => Bytes,
            DbType.Boolean => Bool,
            DbType.Date => Date,
            DbType.DateTime => Timestamp,
            DbType.Double => Float64,
            DbType.Int64 => Int64,
            DbType.VarNumeric => Numeric,
            DbType.Object => Unspecified,
            DbType.String => String,
            _ => throw new ArgumentOutOfRangeException(nameof(DbType), dbType, null),
        };
```

(We could do the same for the DbType property, but I'm not suggesting that change in this PR.)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs:252]: +1

(I suspect this is an inherent problem in the "let's introduce a feature latently, but not make it public yet" - when it *does* go public, it can be hard to make sure we catch everything like this.)

**Review by jskeet — APPROVED** [review]
Okay, this now looks good to me - worth just waiting for a final approval from Amanda as well, in case I've missed any compatibility issues, but I think this one *should* be non-breaking, given that we're only *adding* a mapping rather than changing one, yes?

**Comment by olavloite** [comment]
> Okay, this now looks good to me - worth just waiting for a final approval from Amanda as well, in case I've missed any compatibility issues, but I think this one _should_ be non-breaking, given that we're only _adding_ a mapping rather than changing one, yes?

AFAICT the only 'breaking' change in this case is that if you currently try to only set a `SpannerNumeric` value on a parameter without also setting a type, you will get an error. After this change it will work. We are not changing the behavior of any of the other currently working (or failing) cases.

**Comment by jskeet** [comment]
Forced the CLA bit: for some reason, when a suggestion from me is accepted, the CLA bot doesn't like it :(

**Review by amanda-tarafa — APPROVED** [review]
LGTM. I can't think of any compatibility issues here and this looks good.

**Comment by olavloite** [comment]
@amanda-tarafa Would you mind merging the PR? I don't have permission to merge anything in this repo.

**Comment by amanda-tarafa** [comment]
Ah yes, done!

**Pull request merged**
